File: frontend/src/types/dto/complianceRequirements.ts
/**
 * Реестр требований — форма ответа `GET /compliance/requirements`
 * (backend/app/schemas/compliance_requirements.py; B.18 разд. 19.2, срез-145).
 *
 * `overdue` и `days_left` считает сервер по UTC-«сегодня»: витрина их не
 * пересчитывает, иначе вечером по Москве «сегодня» разошлось бы с сервером.
 * Имена (`owner_name`, `site_name`, `role_label`) тоже приходят с сервера:
 * витрина ничего не знает о пользователях и площадках, кроме того, что ей
 * отдали для выбора.
 */

export type RequirementSeverity = "low" | "medium" | "high" | "critical";
export type RequirementStatus = "active" | "fulfilled" | "retired";

export interface ComplianceEvidenceDto {
  id: string;
  requirement_id: string;
  document_id?: string | null;
  document_title?: string | null;
  note?: string | null;
  confirmed_at: string;
  confirmed_by?: string | null;
  created_at: string;
}

export interface ComplianceRequirementDto {
  id: string;
  code: string;
  title: string;
  description?: string | null;
  npa_id?: string | null;
  npa_code?: string | null;
  npa_title?: string | null;
  clause_id?: string | null;
  clause_code?: string | null;
  role_code?: string | null;
  /** Роль словами (срез-147); код — для машин. */
  role_label?: string | null;
  site_id?: string | null;
  /** Имя площадки (срез-147): без него привязка к объекту невидима. */
  site_name?: string | null;
  process_code?: string | null;
  /** Процесс словами (срез-196). Неизвестный код возвращается как есть: у
   * требований до среза в поле лежит свободный текст. */
  process_label?: string | null;
  owner_user_id?: string | null;
  owner_name?: string | null;
  periodicity_days?: number | null;
  next_due_at?: string | null;
  last_confirmed_at?: string | null;
  severity: RequirementSeverity;
  status: RequirementStatus;
  retired_at?: string | null;
  overdue: boolean;
  days_left?: number | null;
  evidence_count: number;
  created_at: string;
  updated_at: string;
}

export interface ComplianceRequirementDetailDto
  extends ComplianceRequirementDto {
  evidence: ComplianceEvidenceDto[];
}

export interface ComplianceRequirementListDto {
  items: ComplianceRequirementDto[];
  /**
   * Счётчики по ВСЕЙ выборке при текущих фильтрах, а не по выданной странице
   * (срез-195): счётчик по странице заставил бы человека сделать ложный вывод
   * «просроченных нет», увидев «просрочено: 0» над первой страницей.
   */
  total: number;
  active: number;
  overdue: number;
  /** Какая часть выдана. Длина списка этого не заменяет: пустая последняя
   * страница неотличима от «ничего не найдено». */
  limit: number;
  offset: number;
  /** Право заводить, подтверждать и снимать с контроля (admin/owner/ot_specialist). */
  can_manage: boolean;
}

export interface ComplianceRequirementCreateDto {
  code: string;
  title: string;
  description?: string | null;
  npa_id?: string | null;
  clause_id?: string | null;
  role_code?: string | null;
  site_id?: string | null;
  process_code?: string | null;
  owner_user_id?: string | null;
  periodicity_days?: number | null;
  next_due_at?: string | null;
  severity?: RequirementSeverity;
}

/** Правка (срез-147): всё, кроме кода — он естественный ключ арендатора. */
export type ComplianceRequirementUpdateDto = Omit<
  ComplianceRequirementCreateDto,
  "code"
>;

export interface ComplianceEvidenceCreateDto {
  document_id?: string | null;
  note?: string | null;
  confirmed_at?: string | null;
}

export interface ComplianceRequirementFiltersDto {
  npa_id?: string;
  status?: RequirementStatus;
  overdue?: boolean;
  /** Срез-195: реестр листается страницами (по умолчанию сервер отдаёт 50). */
  limit?: number;
  offset?: number;
}

/** Кандидат в ответственные: только то, что нужно, чтобы выбрать человека. */
export interface RequirementOwnerOptionDto {
  id: string;
  name: string;
  role: string;
  role_label: string;
}

export interface RequirementSiteOptionDto {
  id: string;
  name: string;
  /** Компания площадки: у аутсорсера «Цех №1» бывает у нескольких клиентов. */
  company_name: string;
}

export interface RequirementRoleOptionDto {
  code: string;
  label: string;
}

/** Справочники формы требования — `GET /compliance/requirements/options` (срез-147). */
export interface ComplianceRequirementOptionsDto {
  owners: RequirementOwnerOptionDto[];
  sites: RequirementSiteOptionDto[];
  roles: RequirementRoleOptionDto[];
  /** Процессы (срез-196): дисциплины ТЗ + общая охрана труда. */
  processes: RequirementRoleOptionDto[];
}

File: backend/app/schemas/compliance_requirements.py
"""Схемы реестра требований (B.18 разд. 19.2, срез-145)."""

from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field

RequirementSeverity = Literal["low", "medium", "high", "critical"]
RequirementStatus = Literal["active", "fulfilled", "retired"]


class ComplianceRequirementCreate(BaseModel):
    code: str = Field(min_length=1, max_length=64)
    title: str = Field(min_length=1, max_length=255)
    description: str | None = None
    npa_id: str | None = None
    clause_id: str | None = None
    role_code: str | None = Field(default=None, max_length=64)
    site_id: str | None = None
    process_code: str | None = Field(default=None, max_length=64)
    owner_user_id: str | None = None
    #: Дней между исполнениями; пусто — разовое требование.
    periodicity_days: int | None = Field(default=None, ge=1, le=3660)
    next_due_at: date | None = None
    severity: RequirementSeverity = "medium"


class ComplianceRequirementUpdate(BaseModel):
    """Частичное обновление: присланные поля меняются, остальные — нет.

    Код и статус здесь не правятся: код — естественный ключ (ссылки на него
    живут в доказательствах и импорте), статус меняют ручки «исполнено» и
    «снять с контроля», чтобы переход был осмысленным, а не произвольным.
    """

    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    npa_id: str | None = None
    clause_id: str | None = None
    role_code: str | None = Field(default=None, max_length=64)
    site_id: str | None = None
    process_code: str | None = Field(default=None, max_length=64)
    owner_user_id: str | None = None
    periodicity_days: int | None = Field(default=None, ge=1, le=3660)
    next_due_at: date | None = None
    severity: RequirementSeverity | None = None


class ComplianceEvidenceCreate(BaseModel):
    """Доказательство исполнения: документ и/или заметка.

    ``confirmed_at`` — день исполнения (по умолчанию сегодня по UTC); от него
    отсчитывается следующая контрольная дата.
    """

    document_id: str | None = None
    note: str | None = Field(default=None, max_length=2000)
    confirmed_at: date | None = None


class ComplianceEvidenceRead(BaseModel):
    id: str
    requirement_id: str
    document_id: str | None = None
    document_title: str | None = None
    note: str | None = None
    confirmed_at: date
    confirmed_by: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class ComplianceRequirementRead(BaseModel):
    id: str
    code: str
    title: str
    description: str | None = None
    npa_id: str | None = None
    npa_code: str | None = None
    npa_title: str | None = None
    clause_id: str | None = None
    clause_code: str | None = None
    role_code: str | None = None
    #: Роль словами (срез-147, словарь ``app.core.role_labels``); код — для машин.
    role_label: str | None = None
    site_id: str | None = None
    #: Имя площадки (срез-147): без него привязка к объекту на витрине невидима.
    site_name: str | None = None
    process_code: str | None = None
    owner_user_id: str | None = None
    owner_name: str | None = None
    periodicity_days: int | None = None
    next_due_at: date | None = None
    last_confirmed_at: date | None = None
    severity: RequirementSeverity
    status: RequirementStatus
    retired_at: datetime | None = None
    #: Вычислено на чтении по UTC-«сегодня» (срез-140): контрольная дата
    #: прошла, требование ещё активно.
    overdue: bool = False
    #: Дней до контрольной даты (отрицательно — просрочено); None без даты.
    days_left: int | None = None
    evidence_count: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ComplianceRequirementDetail(ComplianceRequirementRead):
    evidence: list[ComplianceEvidenceRead] = Field(default_factory=list)


class ComplianceRequirementListResponse(BaseModel):
    items: list[ComplianceRequirementRead]
    #: Счётчики по ВСЕЙ выборке при текущих фильтрах, а не по выданной странице
    #: (срез-195). Счётчик по странице — класс ошибки, который в проекте ловили
    #: дважды: человек видит «просрочено: 0» над первой страницей и считает,
    #: что просроченных нет вовсе.
    total: int
    active: int
    overdue: int
    #: Может ли пришедший заводить и закрывать требования (роли записи).
    can_manage: bool = False
    #: Какая часть выдана. Клиент не должен вычислять это из длины списка:
    #: пустая последняя страница неотличима от «ничего не найдено».
    limit: int = 0
    offset: int = 0


class RequirementOwnerOption(BaseModel):
    """Кандидат в ответственные: только то, что нужно, чтобы выбрать человека."""

    id: str
    name: str
    role: str
    role_label: str


class RequirementSiteOption(BaseModel):
    id: str
    name: str
    #: Компания площадки: у арендатора-аутсорсера «Цех №1» бывает у нескольких клиентов.
    company_name: str


class RequirementRoleOption(BaseModel):
    code: str
    label: str


class ComplianceRequirementOptions(BaseModel):
    """Справочники формы требования (срез-147): кого назначить ответственным,
    к какой площадке и к какой роли отнести. Отдаются ролям записи."""

    owners: list[RequirementOwnerOption] = Field(default_factory=list)
    sites: list[RequirementSiteOption] = Field(default_factory=list)
    roles: list[RequirementRoleOption] = Field(default_factory=list)
